annotations/relations.py

```diff
@@ -521,3 +522,86 @@ def update_relationset(template, raw_data, creator, text, project_id=None, relat
+    _set_relation_data(template_part, provided, relationset, relation_data, field_handlers, inspect.currentframe().f_code.co_name, cache, appellation_cache, project_id, creator, text)
+    relation = Relation.objects.create(**relation_data)
+
+    if cache != None:
+        cache[template_part.id] = relation
+    return relation
+
+def _update_relation(template_part, provided, relationset, cache={}, appellation_cache={}, project_id=None, creator=None, text=None, required=None, _as_key=None):
+    relation_cache = _get_cache(template_part, cache)
+    if relation_cache != None:
+        return relation_cache
+
+    field_handlers = _get_field_handlers(appellation_cache, project_id, creator, text, required, _as_key)
+    relation_data = _get_relation_data(relationset, creator, text)
+    _set_relation_data(template_part, provided, relationset, relation_data, field_handlers, inspect.currentframe().f_code.co_name, cache, appellation_cache, project_id, creator, text)
+
+    relation_id_q = relationset.constituents.all().values('id')
+    relation = get_object_or_404(Relation, pk=relation_id_q[0]['id'])
+    relation.source_content_object = relation_data['source_content_object']
+    relation.predicate = relation_data['predicate']
+    relation.object_content_object = relation_data['object_content_object']
+    relation.save()
+    
+    if cache != None:
+        cache[template_part.id] = relation
+    return relation
+
+def _set_relation_data(template_part, provided, relationset, relation_data, field_handlers, method_name, cache={}, appellation_cache={}, project_id=None, creator=None, text=None):
+    for pred in ['source', 'predicate', 'object']:    # Collect field data
+        node_type = getattr(template_part, '%s_node_type' % pred)
+        method = field_handlers.get(node_type, field_handlers['__other__'])
+        datum = provided.get((template_part.id, pred))
+
+        dkey = 'predicate' if pred == 'predicate' else '%s_content_object' % pred
+        if datum:
+            relation_data[dkey] = method(datum)
+        elif node_type == RelationTemplatePart.RELATION:
+            relation_data[dkey] = getattr(method_name)(
+                getattr(template_part, PRED_SUFFIX % pred), 
+                provided, 
+                relationset, 
+                cache=cache, 
+                appellation_cache=appellation_cache, 
+                project_id=project_id
+            )
+        else:
+            payload = {
+                'type': node_type,
+                'concept_id': getattr(getattr(template_part, '%s_concept' % pred), 'id', None),
+                'part_field': pred
+            }
+            relation_data[dkey] = create_appellation({}, payload, project_id=project_id, creator=creator, text=text)
+
+def _get_cache(template_part, cache={}):
+    if cache != None:
+        key = template_part.id 
+        if key in cache:
+            return cache[key]
+    return None
```

Comment:
this could be a one liner:
`return cache[template_part.id] if cache and template_part.id in cache else None`

And two other things:
- `_get_cache` implies this method returns a cache. but it returns a cached object (big difference)
- since `template_part` is never used except to get the id, why don't you just pass the id (which would make it easier to read).